algorithms/src/merkle_tree/merkle_tree.rs

```diff
@@ -41,10 +41,30 @@ pub struct MerkleTree<P: MerkleParameters> {
     parameters: Arc<P>,
 }
 
-impl<P: MerkleParameters> MerkleTree<P> {
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+impl<P: MerkleParameters + Send + Sync> MerkleTree<P> {
     pub const DEPTH: u8 = P::DEPTH as u8;
 
-    pub fn new<L: ToBytes, I: ExactSizeIterator<Item = L>>(parameters: Arc<P>, leaves: I) -> Result<Self, MerkleError> {
+    fn hash_row<L: ToBytes + Send + Sync>(
```

Comment:
nit: can you move private methods to the bottom of the impl scope?

Reply:
Is there a philosophy/meaning behind the nit here? I placed it here since it was nearest usage.

Reply:
not sure if its generally used, but in my experience private methods were moved further down as developers using the struct would never call it, so it doesn't serve a meaningful purpose to end-devs looking for methods to call

Reply:
I generally follow the approach of trying to cluster related stuff together, since then it's easier to build a mental model of whats going on with little/no context, without jumping around a lot.